tests/unit/test_repository.py

```diff
@@ -0,0 +1,441 @@
+):
+    """
+    arrange: given Client with an upstream repository with check-branch-exists branch
+    act: when _check_branch_exists is called
+    assert: True is returned.
+    """
+    branch_name = "check-branch-exists"
+    head = upstream_git_repo.create_head(branch_name)
+    head.checkout()
+    (upstream_repository_path / "filler-file").touch()
+    upstream_git_repo.git.add(".")
+    upstream_git_repo.git.commit("-m", "test")
+
+    assert repository_client.check_branch_exists(branch_name)
+
+
+def test_create_branch_error(monkeypatch: pytest.MonkeyPatch, repository_client: Client):
+    """
+    arrange: given Client with a mocked local git repository that raises an exception
+    act: when _create_branch is called
+    assert: RepositoryClientError is raised.
+    """
+    err_str = "mocked error"
+    mock_git_repository = mock.MagicMock(spec=Repo)
+    mock_git_repository.git.commit.side_effect = [GitCommandError(err_str)]
+    monkeypatch.setattr(repository_client, "_git_repo", mock_git_repository)
+
+    with pytest.raises(RepositoryClientError) as exc:
+        repository_client.create_branch(branch_name="test-create-branch", commit_msg="commit-1")
+
+    assert_substrings_in_string(
+        ("unexpected error creating new branch", err_str), str(exc.value).lower()
+    )
+
+
+def test_create_branch(
+    repository_client: Client,
+    repository_path: Path,
+    upstream_git_repo: Repo,
+):
+    """
+    arrange: given Client and newly created files in `repo` and `repo/docs` directories
+    act: when _create_branch is called
+    assert: a new branch is successfully created upstream with only the files in the `repo/docs`
+        directory.
+    """
+    root_file = repository_path / "test.txt"
+    root_file.write_text("content 1", encoding="utf-8")
+    docs_dir = Path(DOCUMENTATION_FOLDER_NAME)
+    (repository_path / docs_dir).mkdir()
+    docs_file = docs_dir / "test.txt"
+    (repository_path / docs_file).write_text("content 2", encoding="utf-8")
+    nested_docs_dir = docs_dir / "nested"
+    (repository_path / nested_docs_dir).mkdir()
+    nested_docs_file = nested_docs_dir / "test.txt"
+    (repository_path / nested_docs_file).write_text("content 3", encoding="utf-8")
+    branch_name = "test-create-branch"
+
+    repository_client.create_branch(branch_name=branch_name, commit_msg="commit-1")
+
+    # mypy false positive in lib due to getter/setter not being next to each other.
+    assert any(
+        branch
+        for branch in upstream_git_repo.branches  # type: ignore
+        if branch.name == branch_name
+    )
+    # Check files in the branch
+    branch_files = set(
+        upstream_git_repo.git.ls_tree("-r", branch_name, "--name-only").splitlines()
+    )
+    assert str(root_file) not in branch_files
+    assert str(docs_file) in branch_files
+    assert str(nested_docs_file) in branch_files
+
+
+def test_create_branch_checkout_clash_default(
```

Comment:
It is because git can confuse the checkout between the file and the branch, if you don't include `--` in the checkout in that case git will actually throw an error. This was actually one of the bugs we found with the migration logic after we first published it, hence the test

Reply:
Ah yes indeed, got it